yarn-project/ethereum/src/l1_tx_utils.test.ts

```diff
@@ -0,0 +1,189 @@
+const WEI_CONST = 1_000_000_000n;
+// Simple contract that just returns 42
+const SIMPLE_CONTRACT_BYTECODE = '0x69602a60005260206000f3600052600a6016f3';
+
+const startAnvil = async (l1BlockTime?: number) => {
+  const ethereumHostPort = await getPort();
+  const rpcUrl = `http://127.0.0.1:${ethereumHostPort}`;
+  const anvil = createAnvil({
+    port: ethereumHostPort,
+    blockTime: l1BlockTime,
+  });
+  await anvil.start();
+  return { anvil, rpcUrl };
+};
+
+describe('GasUtils', () => {
+  let gasUtils: L1TxUtils;
+  let publicClient: any;
+  let walletClient: any;
+  let anvil: Anvil;
+  let initialBaseFee: bigint;
+  const logger = createDebugLogger('l1_gas_test');
+
+  beforeAll(async () => {
+    const { anvil: anvilInstance, rpcUrl } = await startAnvil(1);
+    anvil = anvilInstance;
+    const hdAccount = mnemonicToAccount(MNEMONIC, { addressIndex: 0 });
+    const privKeyRaw = hdAccount.getHdKey().privateKey;
+    if (!privKeyRaw) {
+      throw new Error('Failed to get private key');
+    }
+    const privKey = Buffer.from(privKeyRaw).toString('hex');
+    const account = privateKeyToAccount(`0x${privKey}`);
+
+    publicClient = createPublicClient({
+      transport: http(rpcUrl),
+      chain: foundry,
+    });
+
+    walletClient = createWalletClient({
+      transport: http(rpcUrl),
+      chain: foundry,
+      account,
+    });
+
+    gasUtils = new L1TxUtils(publicClient, walletClient, logger, {
+      bufferPercentage: 20n,
+      maxGwei: 500n,
+      minGwei: 1n,
+      maxAttempts: 3,
+      checkIntervalMs: 100,
+      stallTimeMs: 1000,
+    });
+  });
+  afterAll(async () => {
+    await anvil.stop();
+  }, 5000);
+
+  it('sends and monitors a simple transaction', async () => {
+    const receipt = await gasUtils.sendAndMonitorTransaction({
+      to: '0x1234567890123456789012345678901234567890',
+      data: '0x',
+      value: 0n,
+    });
+
+    expect(receipt.status).toBe('success');
+  }, 10_000);
+
+  it('handles gas price spikes by retrying with higher gas price', async () => {
+    // Get initial base fee
+    const initialBlock = await publicClient.getBlock({ blockTag: 'latest' });
+    initialBaseFee = initialBlock.baseFeePerGas ?? 0n;
+
+    // Start a transaction
+    const sendPromise = gasUtils.sendAndMonitorTransaction({
+      to: '0x1234567890123456789012345678901234567890',
+      data: '0x',
+      value: 0n,
+    });
+
+    // Spike gas price to 3x the initial base fee
+    await publicClient.transport.request({
+      method: 'anvil_setNextBlockBaseFeePerGas',
+      params: [(initialBaseFee * 3n).toString()],
+    });
+
+    // Transaction should still complete
+    const receipt = await sendPromise;
+    expect(receipt.status).toBe('success');
+
+    // Reset base fee
+    await publicClient.transport.request({
+      method: 'anvil_setNextBlockBaseFeePerGas',
+      params: [initialBaseFee.toString()],
+    });
+  });
+
+  it('respects max gas price limits during spikes', async () => {
```

Comment:
I'm worried that neither this nor the previous test assert that there was an actual replacement happening behind the scenes. Can we somehow verify that the tx that was mined is different than the original?

Reply:
you were right, it doesn't issue a new tx. 
We might have to chat about this specifically as I'm not sure how to create this condition with anvil.
If I send a tx with fee too low, I get an error from the RPC directly. 
If I send the tx then manage to increase fee without the tx getting picked, it drops the tx

Reply:
Hmm I think you should be able to disable automining, then use `txpool_inspect` to get the tx in the mempool, and then `anvil_dropTransaction` to manually remove it and force a resubmission. If you don't want to remove the tx from the mempool, maybe you could drop it, mine a block with `evm_mine`, and re-add it real quickly?